test/immutablearray.jl

```diff
@@ -0,0 +1,81 @@
+using Test
+import Core: ImmutableArray, arrayfreeze, mutating_arrayfreeze, arraythaw
+import Core.Compiler: arrayfreeze_tfunc, mutating_arrayfreeze_tfunc, arraythaw_tfunc
+
+@testset "basic ImmutableArray functionality" begin
+    eltypes = (Float16, Float32, Float64, Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64, Int128, UInt128)
+    for t in eltypes
+        a = rand(t, rand(1:100), rand(1:10))
+        b = ImmutableArray(a)
+        @test a == b
+        @test a !== b
+        @test length(a) == length(b)
+        for i in 1:length(a)
+            getindex(a, i) == getindex(b, i)
+        end
+        @test size(a) == size(b)
+        if t in (Float16, Float32, Float64)
+            # @test_broken sum(a) == sum(b) # issue #43772, sometimes works, sometimes doesn't
+        else
+            @test sum(a) == sum(b)
+        end
+        @test reverse(a) == reverse(b)
+        @test ndims(a) == ndims(b)
+        @test axes(a) == axes(b)
+        @test strides(a) == strides(b)
+        @test keys(a) == keys(b)
+        @test_broken IndexStyle(a) == IndexStyle(b) # ImmutableArray is IndexCartesian whereas Array is IndexLinear - worth looking into
```

Comment:
Maybe we want to define this overload for `ImmutableArray`?
https://github.com/ianatol/julia/blob/c4c7acf0f789f29da825b4981b3a50fc9c8ce532/base/indices.jl#L97

I'm not too familiar with `IndexStyle` though. Maybe can we ask a help from @oscardssmith?

Reply:
why not just make `ImmutableArray` have `IndexStyle` `Linear`? I think that would make more sense.

Reply:
Yeah, just hadn’t implemented it yet so its defaulted to Cartesian

Reply:
Since `ImmutableArray` is supposed to be functional as like `Array`, I think it also should use `IndexLinear`.
Though I'm not too familiar with those concepts and wonder why `CartesianIndex` is defined as default, and if there is some requirements for `ImmutableArray` to implement `IndexLinear`.

Reply:
The basic reasoning is that `CartesianIndex` is the less bad one if you get it wrong since cartesian to linear uses multiplication, while linear to cartesian needs division.

Reply:
Cant' it happen for `ImmutableArray`, right? Since `getindex(::ImmutableArray, ...)` is directly forwarded to `arrayref`, I don't think such unsafe indexing operations won't happen as far as it is used usually..